src/webgpu/shader/validation/expression/call/builtin/dot.spec.ts

```diff
@@ -0,0 +1,189 @@
+const builtin = 'dot';
+export const description = `
+Validation tests for the ${builtin}() builtin.
+`;
+
+import { makeTestGroup } from '../../../../../../common/framework/test_group.js';
+import { keysOf, objectsToRecord } from '../../../../../../common/util/data_tables.js';
+import { kValue } from '../../../../../util/constants.js';
+import {
+  Type,
+  kConvertableToFloatVectors,
+  scalarTypeOf,
+  ScalarType,
+} from '../../../../../util/conversion.js';
+import { QuantizeFunc, quantizeToF16, quantizeToF32 } from '../../../../../util/math.js';
+import { ShaderValidationTest } from '../../../shader_validation_test.js';
+
+import {
+  fullRangeForType,
+  kConstantAndOverrideStages,
+  stageSupportsType,
+  validateConstOrOverrideBuiltinEval,
+} from './const_override_validation.js';
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+const kValidArgumentTypes = objectsToRecord(kConvertableToFloatVectors);
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function quantizeFunctionForScalarType(type: ScalarType): QuantizeFunc<any> {
+  switch (type) {
+    case Type.f32:
+      return quantizeToF32;
+    case Type.f16:
+      return quantizeToF16;
+    case Type.abstractInt:
+      return BigInt;
+    default:
+      return Number;
+  }
+}
+
+export interface CheckFunc {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  (num: any): boolean;
+}
+
+function checkFunctionForScalarType(type: ScalarType): CheckFunc {
+  switch (type) {
+    case Type.abstractInt:
+      return (num: bigint) => !kValue.i64.isOOB(num);
+    default:
+      return (num: number) => Number.isFinite(num);
+  }
+}
+
+interface StepArgs {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  [name: string]: any;
+}
+
+export interface EquationStep {
+  (args: StepArgs): StepArgs;
+}
+
+function checkEquation(type: ScalarType, args: StepArgs, steps: EquationStep[]) {
+  const quantizeFn = quantizeFunctionForScalarType(type);
+  const checkFn = checkFunctionForScalarType(type);
+
+  const quantizedArgs: StepArgs = {};
+  for (const key in args) {
+    quantizedArgs[key] = quantizeFn(args[key]);
```

Comment:
This is converting `vecSize` to the scalar type, which is very subtle.

I had a play to see if I could restructure things in a way that avoided `any`, and kept things a bit more type safe. What do you think about this:

```ts

/** Quantizes and validates if the value is representable by the scalar type. */
function quantizeAndCheck<N extends number | bigint>(
  value: N,
  type: ScalarType
): {
  // 'value' quantized to the scalar type.
  quantized: N;
  // true iff the value is representable by the scalar type without error.
  isValid: boolean;
} {
  switch (type) {
    case Type.f32: {
      assert(typeof value === 'number');
      const quantized = quantizeToF32(value) as N;
      return { quantized, isValid: Number.isFinite(quantized) };
    }
    case Type.f16: {
      assert(typeof value === 'number');
      const quantized = quantizeToF16(value) as N;
      return { quantized, isValid: Number.isFinite(quantized) };
    }
    case Type.abstractInt:
      assert(typeof value === 'bigint');
      return { quantized: value, isValid: !kValue.i64.isOOB(value) };
    default:
      assert(typeof value === 'number');
      return { quantized: value, isValid: Number.isFinite(value) };
  }
}

/** Multiplies the two values, coercing to number if the types are not the same */
function mul(a: number | bigint, b: number | bigint): number | bigint {
  if (typeof a === typeof b) {
    return (a as number) * (b as number); // also works for bigint
  }
  // Type mismatch. Coerce to number.
  return Number(a) * Number(b);
}

class Equation<VALUES extends { [key: string]: number | bigint }> {
  /** The dictionary of values */
  readonly values: VALUES;
  /** The scalar type used for equation calculations */
  readonly type: ScalarType;
  /** True if all computed values are valid for type */
  readonly isValid: boolean;

  constructor(type: ScalarType, data: VALUES, isValid?: boolean) {
    this.type = type;
    this.values = data;
    this.isValid = isValid ?? true;
  }

  /**
   * @returns a new Equation extended from this Equation, with an additional field with the name
   * 'name' and value calculated from calling 'fn'. The returned Equation.isValid will be true iff
   * 'this.isValid' is true and the returned value by 'fn' is valid for 'this.type'.
   */
  then<T extends number | bigint, N extends string>(
    name: N,
    fn: (data: VALUES) => T
  ): Equation<VALUES & { [name in N]: T }> {
    const result = fn(this.values);
    const { quantized, isValid } = quantizeAndCheck(result, this.type);
    const out = { ...this.values, [name]: quantized };
    return new Equation(this.type, out, isValid && this.isValid);
  }
}

g.test('values')
  .desc(
    `
Validates that constant evaluation and override evaluation of ${builtin}() never errors
`
  )
  .params(u =>
    u
      .combine('stage', kConstantAndOverrideStages)
      .combine('type', keysOf(kValidArgumentTypes))
      .filter(u => stageSupportsType(u.stage, kValidArgumentTypes[u.type]))
      .beginSubcases()
      .expand('a', u => fullRangeForType(kValidArgumentTypes[u.type], 5))
      .expand('b', u => fullRangeForType(kValidArgumentTypes[u.type], 5))
  )
  .beforeAllSubcases(t => {
    if (scalarTypeOf(kValidArgumentTypes[t.params.type]) === Type.f16) {
      t.selectDeviceOrSkipTestCase('shader-f16');
    }
  })
  .fn(t => {
    const scalarType = scalarTypeOf(kValidArgumentTypes[t.params.type]);

    // Dot product equation: (a[0]*b[0]) + (a[1]*b[1]) + ... (a[N]*b[N])
    // Should be invalid if the dot product calculations result in intermediate
    // values that exceed the maximum representable float value for the given type.

    const expectedResult = new Equation(scalarType, {
      a: t.params.a,
      b: t.params.b,
      vecSize: BigInt(kValidArgumentTypes[t.params.type].width),
    })
      .then('ab', v => mul(v.a, v.b))
      .then('dp', v => mul(v.ab, v.vecSize));

    const type = kValidArgumentTypes[t.params.type];

    // Validates dot(vecN(a), vecN(b));
    validateConstOrOverrideBuiltinEval(
      t,
      builtin,
      expectedResult.isValid,
      [type.create(t.params.a), type.create(t.params.b)],
      t.params.stage
    );
  });
```

Reply:
I appreciate this! Seems viable, but after walking away for a bit and revisiting this I think something more straightforward is probably more prudent here, so now I'm using a simpler utility I put in place for other tests and just doing a conditional, repeating the (short) equation for both bigint and number.